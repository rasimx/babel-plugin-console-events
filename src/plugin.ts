import type babelCore from '@babel/core'
import { types } from "@babel/core";

type VisitorState = {
  file: {
    opts: babelCore.TransformOptions
  }
}

function consoleEventsPlugin(babel: typeof babelCore): babelCore.PluginObj<VisitorState> {
  const {types: t, template} = babel;

  return {
    name: "console-events-plugin",
    visitor: {
      CallExpression(path, state) {

        if (t.isMemberExpression(path.node.callee) &&
          t.isIdentifier(path.node.callee.object) && path.node.callee.object.name == "console" &&
          t.isIdentifier(path.node.callee.property) &&
          ['log', 'info', 'error', 'debug', 'warn'].includes(path.node.callee.property.name) &&
          t.isExpressionStatement(path.parentPath)
        ) {

          const fileName = state.file.opts.filename?.split('/').splice(-1)[0];
          const lineNumber = path.node.loc?.start.line

          const newNode = template(`
                              if (console.events) {
                                console.events.emit({
                                  type: TYPE,
                                  args: ARGS,
                                  loc: LOC,
                                });
                              }
                            `);
          path.insertAfter(newNode({
            TYPE: t.stringLiteral(path.node.callee.property.name),
            ARGS: t.arrayExpression(path.node.arguments as Array<types.Expression>),
            LOC: fileName && lineNumber
              ? t.arrayExpression([t.stringLiteral(fileName), t.stringLiteral(lineNumber.toString())])
              : types.nullLiteral(),
          }));
        }
      }
    }
  }
}

export default consoleEventsPlugin
