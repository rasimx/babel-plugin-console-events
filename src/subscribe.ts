export type ConsoleEvent = {
  type: 'log' | 'info' | 'warn' | 'debug' | 'error',
  args: any[],
  loc: [string, string]
}

export type Listener = (event: ConsoleEvent) => void

class EventEmitter {
  private listeners: Listener[]
  constructor() {
    this.listeners = [];
  }

  emit(event: ConsoleEvent): void {
    this.listeners.forEach((listener) => {
      listener(event)
    })
  }

  subscribe(listener: Listener): Listener {
    this.listeners.push(listener)
    return listener
  }
}

export function subscribe(listener: Listener): void{
  //@ts-ignore
  console.events = console.events ?? new EventEmitter();
  //@ts-ignore
  console.events.subscribe(listener)
}


