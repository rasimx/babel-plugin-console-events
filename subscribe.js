module.exports = require('./lib/subscribe');