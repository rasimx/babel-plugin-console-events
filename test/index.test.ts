import * as path from "path";
import { transformFileSync } from '@babel/core';
import subscribe, { ConsoleEvent } from "../src/subscribe";
import plugin from '../src/plugin'

describe('subscribe', () => {
  it('called with the passed arguments', () => {
    const listenerStub = jest.fn()
    subscribe(listenerStub)

    const event: ConsoleEvent = {
      type: "warn",
      args: ["test string", 123],
      loc: ["code.ts", "5"],
    }

    //@ts-ignore
    console.events.emit(event);
    expect(listenerStub).toHaveBeenCalledWith(event);
  })
})

it('plugin', () => {
  expect(transformFileSync(path.join(__dirname, 'fixtures/plugin.ts'), {plugins: [plugin]})!.code).toMatchSnapshot()
})

it("macros", () => {
  expect(transformFileSync(path.join(__dirname, 'fixtures/macro.ts'))!.code).toMatchSnapshot()
})