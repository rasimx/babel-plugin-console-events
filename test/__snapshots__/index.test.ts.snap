// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros 1`] = `
"console.log(['test']);

if (console.events) {
  console.events.emit({
    type: \\"log\\",
    args: ['test'],
    loc: [\\"macro.ts\\", 2]
  });
}

console.warn(['test']);

if (console.events) {
  console.events.emit({
    type: \\"warn\\",
    args: ['test'],
    loc: [\\"macro.ts\\", 3]
  });
}

console.info(['test']);

if (console.events) {
  console.events.emit({
    type: \\"info\\",
    args: ['test'],
    loc: [\\"macro.ts\\", 4]
  });
}

console.error(['test']);

if (console.events) {
  console.events.emit({
    type: \\"error\\",
    args: ['test'],
    loc: [\\"macro.ts\\", 5]
  });
}

console.debug(['test']);

if (console.events) {
  console.events.emit({
    type: \\"debug\\",
    args: ['test'],
    loc: [\\"macro.ts\\", 6]
  });
}"
`;

exports[`plugin 1`] = `
"// @ts-nocheck
console.log(\\"test string\\", 123);

if (console.events) {
  console.events.emit({
    type: \\"log\\",
    args: [\\"test string\\", 123],
    loc: [\\"plugin.ts\\", \\"2\\"]
  });
}

console.info(\\"test string\\", 123);

if (console.events) {
  console.events.emit({
    type: \\"info\\",
    args: [\\"test string\\", 123],
    loc: [\\"plugin.ts\\", \\"3\\"]
  });
}

console.error(\\"test string\\", 123);

if (console.events) {
  console.events.emit({
    type: \\"error\\",
    args: [\\"test string\\", 123],
    loc: [\\"plugin.ts\\", \\"4\\"]
  });
}

console.warn(\\"test string\\", 123);

if (console.events) {
  console.events.emit({
    type: \\"warn\\",
    args: [\\"test string\\", 123],
    loc: [\\"plugin.ts\\", \\"5\\"]
  });
}

console.debug(\\"test string\\", 123);

if (console.events) {
  console.events.emit({
    type: \\"debug\\",
    args: [\\"test string\\", 123],
    loc: [\\"plugin.ts\\", \\"6\\"]
  });
}

console.test(\\"test string\\", 123); // it shouldn't work here

function f() {}

f(console.log(\\"test string\\", 123)); // prevent stack overflow"
`;
